common/lib/xmodule/xmodule/capa_module.py

```diff
@@ -6,48 +6,122 @@
+from xmodule.x_module import (
+    HTMLSnippet, ResourceTemplates, shim_xmodule_js,
+    XModuleMixin, XModuleToXBlockMixin, XModuleDescriptorToXBlockMixin,
+)
+from xmodule.xml_module import XmlMixin
 
-from .capa_base import CapaFields, CapaMixin, ComplexEncoder
+from .capa_base import _, CapaMixin, ComplexEncoder
 
 log = logging.getLogger("edx.courseware")
 
 
-class CapaModule(CapaMixin, XModule):
+@XBlock.wants('user')  # pylint: disable=abstract-method
+@XBlock.needs('i18n')
+class ProblemBlock(
+        CapaMixin, RawMixin, XmlMixin, EditingMixin,
+        XModuleDescriptorToXBlockMixin, XModuleToXBlockMixin, HTMLSnippet, ResourceTemplates, XModuleMixin):
     """
-    An XModule implementing LonCapa format problems, implemented by way of
-    capa.capa_problem.LoncapaProblem
-
-    CapaModule.__init__ takes the same arguments as xmodule.x_module:XModule.__init__
+    The XBlock for CAPA.
     """
+    INDEX_CONTENT_TYPE = 'CAPA'
+
+    resources_dir = None
+
+    has_score = True
+    show_in_read_only_mode = True
+    template_dir_name = 'problem'
+    mako_template = "widgets/problem-edit.html"
+    has_author_view = True
+
+    # The capa format specifies that what we call max_attempts in the code
+    # is the attribute `attempts`. This will do that conversion
+    metadata_translations = dict(XmlMixin.metadata_translations)
+    metadata_translations['attempts'] = 'max_attempts'
+
     icon_class = 'problem'
 
-    js = {
+    uses_xmodule_styles_setup = True
+    requires_per_student_anonymous_id = True
+
+    preview_view_js = {
         'js': [
             resource_string(__name__, 'js/src/javascript_loader.js'),
             resource_string(__name__, 'js/src/capa/display.js'),
             resource_string(__name__, 'js/src/collapsible.js'),
             resource_string(__name__, 'js/src/capa/imageinput.js'),
             resource_string(__name__, 'js/src/capa/schematic.js'),
+        ],
+        'xmodule_js': resource_string(__name__, 'js/src/xmodule.js')
+    }
+
+    preview_view_css = {
+        'scss': [
+            resource_string(__name__, 'css/capa/display.scss'),
+        ],
+    }
+
+    studio_view_js = {
+        'js': [
+            resource_string(__name__, 'js/src/problem/edit.js'),
+        ],
+        'xmodule_js': resource_string(__name__, 'js/src/xmodule.js'),
+    }
+
+    studio_view_css = {
+        'scss': [
+            resource_string(__name__, 'css/editor/edit.scss'),
+            resource_string(__name__, 'css/problem/edit.scss'),
         ]
     }
 
-    js_module_name = "Problem"
-    css = {'scss': [resource_string(__name__, 'css/capa/display.scss')]}
+    def bind_for_student(self, *args, **kwargs):
+        super(ProblemBlock, self).bind_for_student(*args, **kwargs)
+
+        # Since bind_for_student clears out field_data_cache we also reset self.lcp
+        # which will be regenerated on next access.
+        if 'lcp' in self.__dict__:
+            del self.__dict__['lcp']
```

Comment:
@cpennington I had to do this to fix a few tests:

1. In `openedx.features.content_type_gating.tests.test_access.TestProblemTypeAccess`, `self.score` get reset to `None` when `bind_for_student()` is called again on the block but is not regenerated because `lcp` already exists.
2. There were failures in `lms.djangoapps.instructor_task.tests.test_integration.TestRescoringTask` related to unexpected scores.

Not sure how long Jenkins keeps the test run data but it is at: https://build.testeng.edx.org/job/edx-platform-python-pipeline-pr/3940/#showFailuresLink

Reply:
Sorry, I don't have the stack traces right now. Will get to this tomorrow.

Reply:
When `bind_for_student()` is called, a new `xmodule_runtime` instance is set on the descriptor. Next time a module property is accessed, [this](https://github.com/edx/edx-platform/blob/master/common/lib/xmodule/xmodule/x_module.py#L1207) would be `None` and a new `CapaModule` object gets created.

(I'll improve the docs).

Reply:
@ormsbee I have improved the comment.

`hasattr` executes the property block so that would cause an unnecessary init of the lcp object.